src/Workspaces/Remote/Core/RemoteHostAssetSerialization.cs

```diff
@@ -17,55 +18,93 @@ namespace Microsoft.CodeAnalysis.Remote
-                // pipe for the reader on the other side to read.  This allows the item-writing to remain entirely
-                // synchronous without any blocking on async flushing, while also ensuring that we're not buffering the
-                // entire stream of data into the pipe before it gets sent to the other side.
-                WriteAsset(writer, serializer, context, asset, cancellationToken);
-                await stream.FlushAsync(cancellationToken).ConfigureAwait(false);
+                Contract.ThrowIfNull(asset);
+                foundChecksumCount++;
+
+                using var pooledObject = s_streamPool.GetPooledObject();
+                var tempStream = pooledObject.Object;
+                tempStream.Position = 0;
+
+                // Don't truncate the stream as we're going to be writing to it multiple times.  This will allow us to
+                // reuse the internal chunks of the buffer, without having to reallocate them over and over again.
+                tempStream.SetLength(0, truncate: false);
+
+                // Write the asset to a temporary buffer so we can calculate its length.  Note: as this is an in-memory
+                // temporary buffer, we don't have to worry about synchronous writes on it blocking on the pipe-writer.
+                // Instead, we'll handle the pipe-writing ourselves afterwards in a completely async fasion.
+                WriteAssetToTempStream(tempStream, checksum, asset);
+
+                // Write the length of the asset to the pipe writer so the reader knows how much data to read.
+                WriteLengthToPipeWriter(tempStream.Length);
+
+                // Ensure we flush out the length so the reading side knows how much data to read.
+                await pipeWriterStream.FlushAsync(cancellationToken).ConfigureAwait(false);
+
+                // Now, asynchronously copy the temp buffer over to the writer stream.
+                tempStream.Position = 0;
+                await tempStream.CopyToAsync(pipeWriter, cancellationToken).ConfigureAwait(false);
+
+                // We flush after each item as that forms a reasonably sized chunk of data to want to then send over
+                // the pipe for the reader on the other side to read.  This allows the item-writing to remain
+                // entirely synchronous without any blocking on async flushing, while also ensuring that we're not
+                // buffering the entire stream of data into the pipe before it gets sent to the other side.
+                await pipeWriterStream.FlushAsync(cancellationToken).ConfigureAwait(false);
             }
 
-            return;
+            void WriteAssetToTempStream(Stream tempStream, Checksum checksum, object asset)
+            {
+                using var objectWriter = new ObjectWriter(tempStream, leaveOpen: true, cancellationToken);
```

Comment:
note: ObjectWriter/ObjectReader are a very nice component to use her.e  First off, it provides the nice api that allows assets to richly write themselves out (with int, string, array, etc. support).   Second, the writer writes out a special 2-byte 'version' identifier that the reader looks for.  This allows us to determine immediately if we've screwed something up and we're trying to read at the wrong location in the stream.